Add checks for column constraints in test data for `assure`

milanwiedemann opened the pull request:
This builds on the existing structure that tests the output of ehrQL queries and now also checks if the provided test data meets the constraints. We return details about test data that does not meet the constraints along with the results of testing the ehrQL query instead of returning an error. This makes it possible for users to specify test data that does not meet the constraints. 

Closes #1858

inglesp commented on ehrql/query_model/table_schema.py:
```diff
@@ -22,15 +23,24 @@ def __post_init__(self):
             return f"Possible values: {', '.join(f'`{v}`' for v in self.values)}"
 
+        def validate(self, value):
+            return self.values == value if value else False
```
This is not quite right.  A categorical column is one where values in the column should come from a given tuple (ie `self.values`).  So we should be validating whether the given value is in `self.values`.

inglesp commented on ehrql/query_model/table_schema.py:
```diff
@@ -22,15 +23,24 @@ def __post_init__(self):
+            return value is not None
+
     class Unique(BaseConstraint):
         description = "Always unique"
```
This class could have a `validate` method that returns `True`, with a comment explaining that we can't actually validate a single value.

inglesp commented on ehrql/assurance.py:
```diff
@@ -69,11 +80,31 @@ def validate(variable_definitions, test_data):
+    if len(errors) == 0:
+        return errors
+    else:
+        raise DataValidationError(error_msg)
```
As you said in Slack, raising an exception here means that users would have to fix constraint validation errors one by one.

So instead of raising an exception, why not just return `errors`?  You're already (on lines 66-8) collecting all the constraint validation errors.

The question then becomes what to do with the collected errors.  I should think you could separate out the `present` function into two parts: one that presents any constraint errors, and one that presents any test errors.

milanwiedemann replied:
thx! instead of returning an `DataValidationError`, we have decided to return both (1) all data validation errors and (2) all result validation errors in a message.

milanwiedemann commented on ehrql/assurance.py:
```diff
@@ -69,11 +56,33 @@ def validate(variable_definitions, test_data):
+                    unexpected_test_values.append(
+                        {
+                            "column": column,
+                            "constraint": f"{constraint}",
+                            "value": f"{value}",
```
I should probably use `str()` here instead of f string here for simplicity?

evansd replied:
Yeah, maybe. I think the f-string version is pretty clear though – I wouldn't have picked it out as a thing to change myself.

milanwiedemann replied:
thx, I'll keep it as it is then!

evansd commented on ehrql/query_model/table_schema.py:
```diff
@@ -41,6 +55,9 @@ def __post_init__(self):
             return f"Matches regular expression: `{self.regex}`"
 
+        def validate(self, value):
+            return bool(match(self.regex, value)) if value else True
```
And same again here otherwise the empty string will always pass the test.

evansd reviewed (approved):
This is great, Milan, thanks for doing this.

The only changes that need making are checking for `if value is not None` rather than `if value` – a classic source of bugs in Python.

milanwiedemann commented:
ohh, thanks! that's good to know, I changed `if value` to `if value is not None` where necessary and also added another `assert not c.validate(...)` to the tests.

evansd reviewed (approved):
:+1: All looks good!
